File: launchschool_exercises/109_small_problems/medium2/matching_parentheses.rb
def balanced?(str)
  parens = []
  str.chars do |ch|
    parens << ch if ch == ')' || ch == '('
  end
  
  p parens
  return false if parens.length.odd?
  return false if parens[0] == ')'
  open = 0
  close = 0
  parens.each do |paren|
    if paren == '('
      open += 1
    elsif paren == ')' && close >= open
      return false
    elsif paren == ')'
      close += 1
    end
  end
  true
end

p balanced?('What (is) this?') == true
p balanced?('What is) this?') == false
p balanced?('What (is this?') == false
p balanced?('((What) (is this))?') == true
p balanced?('((What)) (is this))?') == false
p balanced?('Hey!') == true
p balanced?(')Hey!(') == false
p balanced?('What ((is))) up(') == false

File: frontend/210-JavaScript/string_and_text_processing/matching-parentheses.js
function isBalanced(string) {
  var parenOpen = 0;

  for (let i = 0; i < string.length; i++) {
    if (string[i] === '(') {
      parenOpen += 1;
    } else if (string[i] === ')') {
      parenOpen -= 1;
    }  

    if (parenOpen < 0) {
      return false;
    }
  }

  return parenOpen === 0;
}

console.log(isBalanced('What (is) this?'));        // true
console.log(isBalanced('What is) this?'));         // false
console.log(isBalanced('What (is this?'));         // false
console.log(isBalanced('((What) (is this))?'));    // true
console.log(isBalanced('((What)) (is this))?'));   // false
console.log(isBalanced('Hey!'));                   // true
console.log(isBalanced(')Hey!('));                 // false
console.log(isBalanced('What ((is))) up('));       // false
